clients/go/admin/deviceflow/token_orchestrator.go

```diff
@@ -0,0 +1,236 @@
+		return nil, err
+	}
+
+	body, err := io.ReadAll(httpResp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("device authorization request failed due to  %v", err)
+	}
+	if httpResp.StatusCode != http.StatusOK {
+		return nil, &oauth2.RetrieveError{
+			Response: httpResp,
+			Body:     body,
+		}
+	}
+
+	daresp := &DeviceAuthorizationResponse{}
+	err = json.Unmarshal(body, &daresp)
+	if err != nil {
+		return nil, err
+	}
+
+	fmt.Printf("Please open the browser at the url %v and enter following verification code %v\nOR\n", daresp.VerificationURL, daresp.UserCode)
+	fmt.Printf("Please open the browser at the url %v containing verification code\n", daresp.VerificationURLComplete)
+	return daresp, nil
+}
+
+// PollTokenEndpoint polls the token endpoint until the user authorizes/ denies the app or an error occurs other than slow_down or authorization_pending
+func (t TokenOrchestrator) PollTokenEndpoint(ctx context.Context, tokReq DeviceAccessTokenRequest, pollInterval int64) (*oauth2.Token, error) {
+	v := url.Values{
+		cliendID:   {tokReq.ClientID},
+		grantType:  {grantTypeValue},
+		deviceCode: {tokReq.DeviceCode},
+	}
+
+	for {
+
+		httpReq, err := http.NewRequest("POST", t.clientConfig.Endpoint.TokenURL, strings.NewReader(v.Encode()))
+		if err != nil {
+			return nil, err
+		}
+		httpReq.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+
+		httpResp, err := ctxhttp.Do(ctx, nil, httpReq)
+		if err != nil {
+			return nil, err
+		}
+
+		body, err := io.ReadAll(httpResp.Body)
+		if err != nil {
+			return nil, err
+		}
+
+		var tokResp DeviceAccessTokenResponse
+		err = json.Unmarshal(body, &tokResp)
+		if err != nil {
+			return nil, err
+		}
+
+		// Unmarshalled response if it contains an error then check if we need to increase the polling interval
+		if len(tokResp.Error) > 0 {
+			if tokResp.Error == errSlowDown || tokResp.Error == errAuthPending {
+				pollInterval = pollInterval * 2
+
+			} else {
+				return nil, fmt.Errorf("oauth error : %v", tokResp.Error)
+			}
+		} else {
+			// Got the auth token in the response and save it in the cache
+			err = t.tokenCache.SaveToken(&tokResp.Token)
+			// Saving into the cache is only considered to be a warning in this case.
+			if err != nil {
+				logger.Warnf(ctx, "failed to save token in the token cache. Error: %w", err)
+			}
+			return &tokResp.Token, nil
+		}
+		fmt.Printf("Waiting for %v secs\n", pollInterval)
+		time.Sleep(time.Duration(pollInterval) * time.Second)
+	}
+}
+
+// FetchTokenFromAuthFlow starts a webserver to listen to redirect callback from the authorization server at the end
+// of the flow. It then launches the browser to authenticate the user.
+func (t TokenOrchestrator) FetchTokenFromAuthFlow(ctx context.Context) (*oauth2.Token, error) {
+	ctx, cancelNow := context.WithTimeout(ctx, t.cfg.DeviceFlowTimeout.Duration)
+	defer cancelNow()
+
+	var scopes string
+	if len(t.clientConfig.Scopes) > 0 {
+		scopes = strings.Join(t.clientConfig.Scopes, " ")
+	}
```

Comment:
If these are not specified, should we get the scopes from the Metadata endpoint?

Reply:
These are already got from the metadata endpoint. 
We can't send these from cli and is expected to be part of the flyteClientConfig which is returned by the metadata endpoint